src/libclipper/src/selection_policies.cpp

```diff
@@ -1,344 +1,484 @@
+    return selected_model;
+  }
+  double rand_num;
+  srand (time(NULL)); // seed random generator
+  rand_num = (double) rand() / (RAND_MAX); // Pick random number between [0, 1]
+  for (auto it = state.model_map_.begin(); it != state.model_map_.end() && rand_num >= 0; ++it) {
+    rand_num -= (1 - eta) * (it->second["weight"] / state.weight_sum_) + eta / state.model_map_.size();
+    selected_model = it->first;
+  }
+  return selected_model;
 }
 
-std::vector<PredictTask> NewestModelSelectionPolicy::select_predict_tasks(
-    VersionedModelId state, Query query, long query_id) {
-  std::vector<PredictTask> task_vec;
-  // construct the task and put in the vector
-  task_vec.emplace_back(query.input_, state, 1.0, query_id,
-                        query.latency_micros_);
-  return task_vec;
+std::vector<PredictTask> Exp3Policy::select_predict_tasks(PolicyState state,
+                                                Query query,
+                                                long query_id) {
+  auto selected_model = select(state);
+  auto task = PredictTask(query.input_, selected_model, 1.0, query_id,
+                          query.latency_micros_);
+  std::vector<PredictTask> tasks{task};
+  return tasks;
 }
 
-Output NewestModelSelectionPolicy::combine_predictions(
-    VersionedModelId state, Query query, std::vector<Output> predictions) {
-  UNUSED(state);
-  UNUSED(query);
-  // just return the first prediction
+Output Exp3Policy::combine_predictions(PolicyState /*state*/, Query /*query*/,
+                                       std::vector<Output> predictions) {
+  
   if (predictions.empty()) {
-    return Output{0.0, std::make_pair("none", 0)};
-  } else {
-    return predictions.front();
+    std::cout << "No predictions to combine" << std::endl;
+    Output output;
+    return output;
   }
+  return predictions.front();
 }
 
 std::pair<std::vector<PredictTask>, std::vector<FeedbackTask>>
-NewestModelSelectionPolicy::select_feedback_tasks(VersionedModelId state,
-                                                  FeedbackQuery query,
-                                                  long query_id) {
-  UNUSED(state);
-  UNUSED(query);
-  UNUSED(query_id);
-  return std::make_pair(std::vector<PredictTask>(),
-                        std::vector<FeedbackTask>());
-}
-
-VersionedModelId NewestModelSelectionPolicy::process_feedback(
-    VersionedModelId state, Feedback feedback,
-    std::vector<Output> predictions) {
-  UNUSED(feedback);
-  UNUSED(predictions);
+Exp3Policy::select_feedback_tasks(PolicyState& state, FeedbackQuery feedback,
+                                  long query_id) {
+  // Predict Task
+  auto selected_model = select(state);
+  auto predict_task = PredictTask(feedback.feedback_.input_, selected_model, -1, query_id, -1);
+  std::vector<PredictTask> predict_tasks{predict_task};
+  // Feedback Task
+  auto feedback_task = FeedbackTask(feedback.feedback_, selected_model, query_id, -1);
```

Comment:
Should I make select_feedback_tasks function only return std::vector<PredictTask> instead of std::pair<std::vector<PredictTask>, std::vector<FeedbackTask>>?